### PR TITLE
Product Class in Core

### PR DESCRIPTION
**Pull request by tavisca-nnarkhede**
Added Function for getting products and  get a particular product

### CODE REVIEW

File: src/OnlineRetailPortal.Contracts/Contracts/IProductService.cs [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
+﻿using OnlineRetailPortal.Contracts.Models;
+using System;
 using System.Collections.Generic;
 using System.Text;
+using System.Threading.Tasks;
 
 namespace OnlineRetailPortal.Contracts.Contracts
 {
-    interface IProductService
+    public interface IProductService
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Contracts/Contracts/IProductService.cs]: remove the changes

File: src/OnlineRetailPortal.Services/Services/ProductService.cs [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
+﻿using OnlineRetailPortal.Contracts.Models;
+using System;
 using System.Collections.Generic;
 using System.Text;
+using System.Threading.Tasks;
 
 namespace OnlineRetailPortal.Services.Services
 {
-    class ProductService
+    public class ProductService
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Services/Services/ProductService.cs]: remove the changes

Comment (tavisca-nnarkhede) [reply src/OnlineRetailPortal.Services/Services/ProductService.cs]: done

File: src/OnlineRetailPortal.Contracts/Contracts/IProductService.cs (hunk 1 above)
Comment (tavisca-nnarkhede) [reply src/OnlineRetailPortal.Contracts/Contracts/IProductService.cs]: done

File: src/OnlineRetailPortal.Contracts/Contracts/IProductStore.cs [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
+using System;
 using System.Collections.Generic;
 using System.Text;
 
 namespace OnlineRetailPortal.Contracts.Contracts
 {
-    interface IProductStore
+    public interface IProductStore
     {
+        List<Product> getProducts(int pagenumber, int pagesize);
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Contracts/Contracts/IProductStore.cs]: 1. contracts should support async.
2. it should take sort by as well and all these parameters should be part of one class.
Refrain from passing too many arguments to a public method

File: src/OnlineRetailPortal.Core/Models/Price.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OnlineRetailPortal.Core.Models
+{
+    public class Price
+    {
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/Price.cs:8]: it should accept the currency as well

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 5]
```diff
@@ -1,10 +1,50 @@
...
+        public string Description { get; set; }
+        public Status Status { get; set; }
+
+
+        
+        public static async Task<List<Product>> GetAllProducts(int pagenumber, int pagesize)
+        {
+            List<Product> products = new List<Product>();
+            //products = idbInterface.getProducts(pagenumber,pagesize);// This data will be fetched from database
+            if (products.Count == 0)
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/Product.cs:48]: this type of exceptions should not be thrown by library

File: src/OnlineRetailPortal.Core/Models/User.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OnlineRetailPortal.Core.Models
+{
+    class User
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/User.cs]: where are we using this ?

File: src/OnlineRetailPortal.Core/Models/Category.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OnlineRetailPortal.Core.Models
+{
+    public enum Category
+    {
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/Category.cs:8]: I can see abhishek has also declared category refrain from doing repeatative check-ins

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 8]
```diff
@@ -1,10 +1,50 @@
...
+                throw new Exception("No Products Available");
+            }
+            return products;
+        }
+
+        public static async Task<Product> GetProductById(string productId)
+        {
+            Product product = null;
+            //product = idbInterface.getProduct(productId);// This data will be fetched from database
+            if (product == null)
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/Product.cs]: if you want to throw exceptions , make sure library (core throws it's custom exceptions)

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 9]
```diff
@@ -1,10 +1,50 @@
...
+using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Text;
+using System.Threading.Tasks;
 
 namespace OnlineRetailPortal.Core.Models
 {
-    class Product
+    public class Product
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Core/Models/Product.cs:10]: core models could have their own restrictions ( constructors accepting some parameters as mandatory )
please revisit your models

**Review by Tavisca-smohite — CHANGES_REQUESTED** [review]
Please add test cases for your impl

File: src/OnlineRetailPortal.Contracts/Models/Category.cs [hunk 10]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OnlineRetailPortal.Contracts
+{
+    public enum Category
+    {
+        Properties,
```
Comment (tavisca-snikam) [inline src/OnlineRetailPortal.Contracts/Models/Category.cs]: Category should not have Plurals

File: src/OnlineRetailPortal.Core/Models/Page.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OnlineRetailPortal.Core
+{
+    public class Page
```
Comment (tavisca-snikam) [inline src/OnlineRetailPortal.Core/Models/Page.cs]: PagingInfo name makes more sense

File: src/OnlineRetailPortal.Contracts/Models/PageSortBy.cs [hunk 12]
```diff
@@ -4,7 +4,9 @@
 
 namespace OnlineRetailPortal.Contracts
 {
-    class ServiceRequest
+    public enum PageSortBy
```
Comment (tavisca-snikam) [inline src/OnlineRetailPortal.Contracts/Models/PageSortBy.cs:7]: Please remmove Page prefix

File: src/OnlineRetailPortal.Contracts/Models/Price.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OnlineRetailPortal.Contracts
+{
+    public class Price
+    {
+        public double Amount { get; set; }
+        public bool IsNegotiable { get; set; }
```
Comment (tavisca-snikam) [inline src/OnlineRetailPortal.Contracts/Models/Price.cs:10]: I think this property should be a part of Ad.

File: src/OnlineRetailPortal.Contracts/Models/Product.cs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+        public string Name { get; set; }
+        public Price Price { get; set; }
+        public Category Category { get; set; }
+        public Image HeroImage { get; set; }
+        public Address PickupAddress { get; set; }
+        public List<Image> Images { get; set; }
+        public string UserId { get; set; }
+        public DateTime PostDateTime { get; set; }
+        public DateTime PurchasedDate { get; set; }
+        public DateTime ExpirationDate { get; set; }
```
Comment (tavisca-snikam) [inline src/OnlineRetailPortal.Contracts/Models/Product.cs:21]: How are we going to handle expiration. Instead we can have IsActive flag in Ad/Product

File: src/OnlineRetailPortal.Contracts/Models/ProductList.cs [hunk 15]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OnlineRetailPortal.Contracts
+{
+
+        public class ProductList
```
Comment (tavisca-snikam) [inline src/OnlineRetailPortal.Contracts/Models/ProductList.cs:8]: List word is confusing. This represents a data entity not list

Comment (tavisca-snikam) [reply src/OnlineRetailPortal.Contracts/Models/ProductList.cs:8]: Why does this class exist if Product class is already there

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 16]
```diff
@@ -1,25 +1,70 @@
...
+namespace OnlineRetailPortal.Core
 {
     public class Product
     {
         public string Id { get; set; }
         public string Name { get; set; }
-        public string Description { get; set; }
-        public Image HeroImage { get; set; }
         public Price Price { get; set; }
+        public Price IsNegotiable { get; set; }
```
Comment (Swinal-Tavisca) [inline src/OnlineRetailPortal.Core/Models/Product.cs]: IsNegotiable should not be of type Price. verify this

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 17]
```diff
@@ -1,25 +1,70 @@
...
+        {
+            this.productStore = productStore;
+        }
+
+        public async Task<GetProductsServiceResponse> GetProductsAsync(GetProductsServiceRequest serviceRequest)
+        {
+            var request = serviceRequest.ToProductStoreRequest();
+            List<Product> products = new List<Product>();
+            GetProductsStoreResponse response = new GetProductsStoreResponse();
+            response = await productStore.GetProductsAsync(request);
```
Comment (tavisca-snikam) [inline src/OnlineRetailPortal.Core/Models/Product.cs:46]: is it like get all products without pagination

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 18]
```diff
@@ -1,25 +1,70 @@
...
 {
     public class Product
     {
         public string Id { get; set; }
         public string Name { get; set; }
-        public string Description { get; set; }
-        public Image HeroImage { get; set; }
         public Price Price { get; set; }
+        public Price IsNegotiable { get; set; }
+        public Price Currency { get; set; }
```
Comment (Swinal-Tavisca) [inline src/OnlineRetailPortal.Core/Models/Product.cs]: Currency can not be of type Price.

File: src/OnlineRetailPortal.Core/Models/Page.cs (hunk 11 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/Page.cs]: done

File: src/OnlineRetailPortal.Core/Models/Product.cs (hunk 17 above)
Comment (tavisca-snikam) [reply src/OnlineRetailPortal.Core/Models/Product.cs]: I see paging info from request is not considered yet

File: src/OnlineRetailPortal.Core/Models/Product.cs [hunk 19]
```diff
@@ -1,25 +1,70 @@
...
         public DateTime ExpirationDate { get; set; }
 
-        public List<Image> Images { get; set; }
-        public DateTime PurchasedDate { get; set; }
-        public Address PickupAddress { get; set; }
+        public Status Status { get; set; }
+
+        IProductStore productStore;
+
+        public Product()
```
Comment (Swinal-Tavisca) [inline src/OnlineRetailPortal.Core/Models/Product.cs:31]: remove this if its not having any  references

**Review by Swinal-Tavisca — CHANGES_REQUESTED** [review]

File: src/OnlineRetailPortal.Core/Models/ProductList.cs [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+using System.Text;
+
+namespace OnlineRetailPortal.Core
+{
+    public class ProductList
+    {
+        public string Id { get; set; }
+        public string Name { get; set; }
+        public Price Price { get; set; }
+        public Price Currency { get; set; }
```
Comment (Swinal-Tavisca) [inline src/OnlineRetailPortal.Core/Models/ProductList.cs]: again Currency can not be of type Price.

File: src/OnlineRetailPortal.Contracts/Models/ProductList.cs (hunk 15 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Contracts/Models/ProductList.cs:8]: PageList is used for homepage user get a list of products it does not shows all product detail it only shows necessary details... in this pr, I am not using it anywhere.

File: src/OnlineRetailPortal.Core/Translators/GetProductsServiceResponseTranslator.cs [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            var getProductsResponse = new GetProductsServiceResponse()
+            {
+                Products = GetList(products)
+            };
+            return getProductsResponse;
+
+        }
+
+        private static List<Contracts.Product> GetList(List<Product> products)
```
Comment (Swinal-Tavisca) [inline src/OnlineRetailPortal.Core/Translators/GetProductsServiceResponseTranslator.cs]: this is giving products list can rename it as GetProductList

**Review by Swinal-Tavisca — CHANGES_REQUESTED** [review]

File: src/OnlineRetailPortal.Core/Models/Product.cs (hunk 17 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/Product.cs:46]: the request contains paging information.

File: src/OnlineRetailPortal.Contracts/Models/PageSortBy.cs (hunk 12 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Contracts/Models/PageSortBy.cs:7]: done

File: src/OnlineRetailPortal.Core/Models/Product.cs (hunk 17 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/Product.cs:46]: done

File: src/OnlineRetailPortal.Contracts/Models/Product.cs (hunk 14 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Contracts/Models/Product.cs:21]: product has a property status, so we can mark product active, sold or disabled.

File: src/OnlineRetailPortal.Core/Models/ProductList.cs (hunk 20 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/ProductList.cs]: currency is a property of Price, and this property is type of Currency.

File: src/OnlineRetailPortal.Core/Translators/GetProductsServiceResponseTranslator.cs (hunk 21 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Translators/GetProductsServiceResponseTranslator.cs]: done

File: src/OnlineRetailPortal.Core/Models/Product.cs (hunk 16 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/Product.cs]: done

File: src/OnlineRetailPortal.Core/Models/Product.cs (hunk 18 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/Product.cs]: done

File: src/OnlineRetailPortal.Core/Models/Product.cs (hunk 19 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/Product.cs:31]: it is for mock db

File: src/OnlineRetailPortal.Contracts/Models/Currency.cs [hunk 22]
```diff
@@ -4,7 +4,10 @@
 
 namespace OnlineRetailPortal.Contracts
 {
-    class ServiceResponce
+    public enum Currency
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Contracts/Models/Currency.cs]: Can you explain why currency is enum?

File: src/OnlineRetailPortal.Contracts/Models/GetProductsServiceRequest.cs [hunk 23]
```diff
@@ -0,0 +1,7 @@
+﻿namespace OnlineRetailPortal.Contracts
+{
+    public class GetProductsServiceRequest
+    {
+        public PagingInfo Page { get; set; }
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Contracts/Models/GetProductsServiceRequest.cs:5]: The prop name should be pagingInfo

File: src/OnlineRetailPortal.Contracts/Contracts/IProductStore.cs [hunk 24]
```diff
@@ -1,10 +1,15 @@
...
 using System.Collections.Generic;
 using System.Text;
+using System.Threading.Tasks;
 
 namespace OnlineRetailPortal.Contracts
 {
     public interface IProductStore
     {
+        Task<GetProductStoreResponse> GetProductAsync(string productId);
+        Task<GetProductsStoreResponse> GetProductsAsync(GetProductsStoreRequest request);
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Contracts/Contracts/IProductStore.cs:13]: What the other method do?

File: src/OnlineRetailPortal.Contracts/Models/GetProductServiceResponse.cs [hunk 25]
```diff
@@ -0,0 +1,7 @@
+﻿namespace OnlineRetailPortal.Contracts
+{
+    public class GetProductServiceResponse
+    {
+        public Product Product { get; set; }
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Contracts/Models/GetProductServiceResponse.cs:5]: Is the sellerId is Product class’s property?

File: src/OnlineRetailPortal.Contracts/Models/GetProductsServiceResponse.cs [hunk 26]
```diff
@@ -0,0 +1,9 @@
+﻿using System.Collections.Generic;
+
+namespace OnlineRetailPortal.Contracts
+{
+    public class GetProductsServiceResponse
+    {
+        public List<Product> Products { get; set; }
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Contracts/Models/GetProductsServiceResponse.cs:7]: How are we handling pagingInfo in response?
Are you handling the bundling of pagingInfo web response?

File: src/OnlineRetailPortal.Core/Models/Product.cs (hunk 19 above)
Comment (Tavisca-smohite) [reply src/OnlineRetailPortal.Core/Models/Product.cs:31]: You should not add the extra properties or constructors just to fulfil need of any particular impl..
It’s a core class . Your mock impl should handle that on it’s own

File: src/OnlineRetailPortal.Core/Models/User.cs (hunk 6 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Core/Models/User.cs]: done removed

File: src/OnlineRetailPortal.Contracts/Models/ProductList.cs (hunk 15 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Contracts/Models/ProductList.cs:8]: removed

File: src/OnlineRetailPortal.Contracts/Models/GetProductsServiceRequest.cs (hunk 23 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Contracts/Models/GetProductsServiceRequest.cs:5]: done

File: src/OnlineRetailPortal.Contracts/Contracts/IProductStore.cs (hunk 24 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Contracts/Contracts/IProductStore.cs:13]: done

File: src/OnlineRetailPortal.Contracts/Models/GetProductServiceResponse.cs (hunk 25 above)
Comment (tavisca-sgogawale) [reply src/OnlineRetailPortal.Contracts/Models/GetProductServiceResponse.cs:5]: sellerid added

**Pull request closed without merging**